### PR TITLE
Kotlin client/spacings indentations

### PR DESCRIPTION
**Pull request by noordawod**
There has been changes in the Kotlin template that wrecked the beautifully-generated code that I'm used to, plus removed few important utility functions for enums for no obvious reason.

Well, I brought these utility functions back and also added plenty of reserved keywords and operators that were missing from the Kotlin generator.

### PR checklist
 
- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [X] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [X] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@jimschubert @dr4ke616 @karismann @Zomzog @andrewemery @4brunu @yutaka0m

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/kotlin-client/data_class_req_var.mustache [hunk 1]
```diff
@@ -18,4 +15,4 @@
...
\ No newline at end of file
+    {{#multiplatform}}@SerialName(value = "{{{vendorExtensions.x-base-name-literal}}}") @Required {{/multiplatform}}{{#isInherited}}override {{/isInherited}}{{>modelMutable}} `{{name}}`: {{#isArray}}{{#isList}}kotlin.collections.List{{/isList}}{{^isList}}kotlin.Array{{/isList}}<{{^items.isEnum}}{{^items.isPrimitiveType}}{{^items.isModel}}{{#kotlinx_serialization}}@Contextual {{/kotlinx_serialization}}{{/items.isModel}}{{/items.isPrimitiveType}}{{{items.dataType}}}{{/items.isEnum}}{{#items.isEnum}}{{classname}}.{{{nameInCamelCase}}}{{/items.isEnum}}>{{/isArray}}{{^isEnum}}{{^isArray}}{{{dataType}}}{{/isArray}}{{/isEnum}}{{#isEnum}}{{^isArray}}{{classname}}.{{{nameInCamelCase}}}{{/isArray}}{{/isEnum}}{{#isNullable}}?{{/isNullable}}{{#defaultValue}} = {{{defaultValue}}}{{/defaultValue}}
```
Comment (noordawod) [inline modules/openapi-generator/src/main/resources/kotlin-client/data_class_req_var.mustache:18]: Someone mistakenly removed default values for this data class. Brought back now.

**Comment by noordawod** [comment]
The failed tests don't seem to be caused by the changes..

File: modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache [hunk 2]
```diff
@@ -18,4 +15,4 @@
...
\ No newline at end of file
+    {{#multiplatform}}@SerialName(value = "{{{vendorExtensions.x-base-name-literal}}}") {{/multiplatform}}{{#isInherited}}override {{/isInherited}}{{>modelMutable}} `{{name}}`: {{#isArray}}{{#isList}}kotlin.collections.List{{/isList}}{{^isList}}kotlin.Array{{/isList}}<{{^items.isEnum}}{{^items.isPrimitiveType}}{{^items.isModel}}{{#kotlinx_serialization}}@Contextual {{/kotlinx_serialization}}{{/items.isModel}}{{/items.isPrimitiveType}}{{{items.dataType}}}{{/items.isEnum}}{{#items.isEnum}}{{classname}}.{{{nameInCamelCase}}}{{/items.isEnum}}>{{/isArray}}{{^isEnum}}{{^isArray}}{{{dataType}}}{{/isArray}}{{/isEnum}}{{#isEnum}}{{^isArray}}{{classname}}.{{{nameInCamelCase}}}{{/isArray}}{{/isEnum}}? = {{#defaultvalue}}{{{defaultvalue}}}{{/defaultvalue}}{{^defaultvalue}}null{{/defaultvalue}}
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache:18]: @noordawod I'm not sure about this change. I would like to hear someone else on this change.

File: modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache [hunk 3]
```diff
@@ -77,20 +81,20 @@ import java.io.Serializable
     {{#enumVars}}
         {{^multiplatform}}
         {{#moshi}}
-        @Json(name = {{^isString}}"{{/isString}}{{{value}}}{{^isString}}"{{/isString}}) {{&name}}({{{value}}}){{^-last}},{{/-last}}{{#-last}};{{/-last}}
+        @Json(name = {{^isString}}"{{/isString}}{{{value}}}{{^isString}}"{{/isString}}) `{{&name}}`({{{value}}}){{^-last}},{{/-last}}{{#-last}};{{/-last}}
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:84]: Same here

File: modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache (hunk 2 above)
Comment (noordawod) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache:18]: If you have a property which is a reserved word in Kotlin (`internal`, `public`, `when`, etc.) then it won't be escaped automatically by openapi generator. I don't know why, but it just doesn't. So the simplest solution is one of two: 

1) Add an underscore to the end of each property (I believe it was like so in an earlier incarnation).
2) Surround with a backtick.
3) Fix the generator to escape reserved words based on language.

Am I missing something?

File: modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache (hunk 3 above)
Comment (noordawod) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:84]: Yup, yup, and the same reasons apply :)

File: modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache (hunk 2 above)
Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache:18]: This is related to https://github.com/OpenAPITools/openapi-generator/pull/9842
Do you miss any more words?

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache:18]: `public`, `private` were not escaped.

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache:18]: Could you please add those words to the reserved words list, and revert the back-ticks changes?

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache:18]: I can do that. So basically, to touch the Java code, that's what you mean? (`modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractKotlinCodegen.java`)

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache:18]: Yes please 👍

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class_opt_var.mustache]: Done!

File: samples/client/petstore/kotlin-gson/src/main/kotlin/org/openapitools/client/models/Order.kt [hunk 4]
```diff
@@ -25,23 +35,30 @@ import com.google.gson.annotations.SerializedName
...
+
     @SerializedName("petId")
     val petId: kotlin.Long? = null,
+
     @SerializedName("quantity")
     val quantity: kotlin.Int? = null,
+
     @SerializedName("shipDate")
     val shipDate: java.time.OffsetDateTime? = null,
-    /* Order Status */
```
Comment (4brunu) [inline samples/client/petstore/kotlin-gson/src/main/kotlin/org/openapitools/client/models/Order.kt:36]: @noordawod is this change intentional?

File: samples/server/petstore/kotlin-springboot-delegate/src/main/kotlin/org/openapitools/api/PetApi.kt [hunk 5]
```diff
@@ -85,7 +85,7 @@ interface PetApi {
         authorizations = [Authorization(value = "petstore_auth", scopes = [AuthorizationScope(scope = "read:pets", description = "read your pets")])])
     @ApiResponses(
         value = [ApiResponse(code = 200, message = "successful operation", response = Pet::class, responseContainer = "List"),ApiResponse(code = 400, message = "Invalid status value")])
-    @GetMapping(
+    @`get`Mapping(
```
Comment (4brunu) [inline samples/server/petstore/kotlin-springboot-delegate/src/main/kotlin/org/openapitools/api/PetApi.kt:88]: @noordawod is this suppose to happen?

File: samples/server/petstore/kotlin-server/ktor/src/main/kotlin/org/openapitools/server/apis/UserApi.kt [hunk 6]
```diff
@@ -63,11 +63,11 @@ fun Route.UserApi() {
         }
     }
 
-    get<Paths.loginUser> {
+    `get`<Paths.loginUser> {
```
Comment (4brunu) [inline samples/server/petstore/kotlin-server/ktor/src/main/kotlin/org/openapitools/server/apis/UserApi.kt:66]: @noordawod is this suppose to happen?

**Review by 4brunu — COMMENTED** [review]
@noordawod I left some questions to you in the code review, could

File: samples/server/petstore/kotlin-springboot-delegate/src/main/kotlin/org/openapitools/api/PetApi.kt (hunk 5 above)
Comment (noordawod) [reply samples/server/petstore/kotlin-springboot-delegate/src/main/kotlin/org/openapitools/api/PetApi.kt:88]: Hmm, nope... I guess a bug in the generator, so the easiest fix is to remove `get` from the list.

File: samples/client/petstore/kotlin-gson/src/main/kotlin/org/openapitools/client/models/Order.kt (hunk 4 above)
Comment (noordawod) [reply samples/client/petstore/kotlin-gson/src/main/kotlin/org/openapitools/client/models/Order.kt:36]: Nope. I'll check why it happened.

File: samples/server/petstore/kotlin-springboot-delegate/src/main/kotlin/org/openapitools/api/PetApi.kt (hunk 5 above)
Comment (4brunu) [reply samples/server/petstore/kotlin-springboot-delegate/src/main/kotlin/org/openapitools/api/PetApi.kt:88]: I think this already happen once and the solution was to remove `get`.
OR you can add only the words that you need.

**Comment by noordawod** [comment]
Build failed for another reason: `Could not resolve dependencies for project com.prokarma:pkmst-microservice:jar:1.0.0: Could not find artifact io.springfox:springfox-swagger2:jar:2.6.1-SNAPSHOT`

**Comment by noordawod** [comment]
> @noordawod I left some questions to you in the code review, could you please take a look at those? 🙂

I believe all is taken care of now, @4brunu.

**Comment by 4brunu** [comment]
@noordawod Could you please solve the merge conflicts?

File: modules/openapi-generator/src/main/resources/kotlin-client/enum_class.mustache [hunk 7]
```diff
@@ -52,12 +52,29 @@ import kotlinx.serialization.internal.CommonEnumSerializer
...
+     * This solves a problem when the variable name and its value are different, and ensures that
+     * the client sends the correct enum values to the server always.
+     */
+    override fun toString(): String = value{{^isString}}.toString(){{/isString}}
+
+    companion object {
+        /**
+         * Converts the provided [data] to a [String] on success, null otherwise.
+         */
+        fun encode(data: Any?): kotlin.String? = if (data is {{classname}}) "$data" else null
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/enum_class.mustache:68]: @noordawod could you please explain what is the use case for this utility?

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/kotlin-client/enum_class.mustache:68]: Unused by me, but because `decode()` is introduced I wanted to compliment it with an `encode()`, I think it's a good practice.

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/kotlin-client/enum_class.mustache:68]: ps: I use `decode()`.

**Comment by 4brunu** [comment]
@noordawod When you have some time, could you please solve the merge conflicts? 🙂

**Comment by noordawod** [comment]
> @noordawod When you have some time, could you please solve the merge conflicts?

Yup, was on a holiday ;) I'll attend to this asap!

**Comment by 4brunu** [comment]
Sorry, when you have time :)

**Comment by noordawod** [comment]
> Sorry, when you have time :)

Synced and pushed now, let's hope everything passes.

**Comment by noordawod** [comment]
> Sorry, when you have time :)

Some tests aren't passing because of `Could not find org.openapitools:openapi-generator-gradle-plugin:5.2.0-SNAPSHOT`.

**Comment by 4brunu** [comment]
Looks good to me 👍
